kura/org.eclipse.kura.util/src/main/java/org/eclipse/kura/util/service/ServiceSupplier.java

```diff
@@ -0,0 +1,138 @@
+ *
+ * @param <T>
+ *            the target service object
+ */
+public final class ServiceSupplier<T> implements Supplier<List<T>>, AutoCloseable {
+
+    /** Logger instance */
+    private static final Logger logger = LoggerFactory.getLogger(ServiceSupplier.class);
+
+    /** Localization instance */
+    private static final UtilMessages message = LocalizationAdapter.adapt(UtilMessages.class);
+
+    /** Associated Bundle Context */
+    private final BundleContext bundleContext;
+
+    /** Associated Service References */
+    private final Collection<ServiceReference<T>> serviceReferences;
+
+    /** Acquired Service Instances */
+    private final Map<T, ServiceReference<T>> acquiredServices;
+
+    /** Constructor */
+    private ServiceSupplier(final Class<T> target, @Nullable final String filter) {
+        requireNonNull(target, message.targetNonNull());
+        this.bundleContext = ServiceUtil.getBundleContext();
+        this.serviceReferences = ServiceUtil.getServiceReferences(target, filter);
+        this.acquiredServices = CollectionUtil.newHashMap();
+    }
+
+    /** Constructor */
+    private ServiceSupplier(final ServiceReference<T> reference) {
+        requireNonNull(reference, message.referenceNonNull());
+        this.bundleContext = ServiceUtil.getBundleContext();
+        this.serviceReferences = CollectionUtil.newArrayList();
+        this.serviceReferences.add(reference);
+        this.acquiredServices = CollectionUtil.newHashMap();
+    }
+
+    /**
+     * Supplies the instance of {@link ServiceSupplier} from which the target service instance can be
+     * retrieved
+     *
+     * @param target
+     *            the service instance to retrieve
+     * @param filter
+     *            the valid OSGi service filter (can be {@code null})
+     * @return the {@link ServiceSupplier} instance
+     * @throws NullPointerException
+     *             if any of the arguments is {@code null} (except {@code filter})
+     * @throws IllegalArgumentException
+     *             if {@link BundleContext} could not be retrieved
+     */
+    public static <T> ServiceSupplier<T> supply(final Class<T> target, @Nullable final String filter) {
+        return new ServiceSupplier<>(target, filter);
+    }
+
+    /**
+     * Supplies the instance of {@link ServiceSupplier} from which the target service instance can be
+     * retrieved
+     *
+     * @param reference
+     *            the {@link ServiceReference} instance to get the target service
+     * @return the {@link ServiceSupplier} instance
+     * @throws NullPointerException
+     *             if any of the arguments is {@code null}
+     */
+    public static <T> ServiceSupplier<T> supply(final ServiceReference<T> reference) {
+        return new ServiceSupplier<>(reference);
+    }
+
+    /**
+     * Retrieves the list of target service instances
+     *
+     * @return the list of service instances or empty list
+     */
+    @Override
+    public List<T> get() {
```

Comment:
You are right. I have just added a check to validate if the service instance has already been retrieved from the service reference.